plugins/UM3NetworkPrinting/src/Cloud/CloudOutputDevice.py

```diff
@@ -0,0 +1,483 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+import io
+import json
+import os
+from json import JSONDecodeError
+from typing import List, Optional, Dict, cast, Union, Tuple
+
+from PyQt5.QtCore import QObject, pyqtSignal, QUrl, pyqtProperty, pyqtSlot
+from PyQt5.QtNetwork import QNetworkReply, QNetworkRequest
+
+from UM import i18nCatalog
+from UM.FileHandler.FileWriter import FileWriter
+from UM.FileHandler.FileHandler import FileHandler
+from UM.Logger import Logger
+from UM.Message import Message
+from UM.OutputDevice import OutputDeviceError
+from UM.Scene.SceneNode import SceneNode
+from UM.Version import Version
+from cura.CuraApplication import CuraApplication
+from cura.PrinterOutput.PrinterOutputController import PrinterOutputController
+from cura.PrinterOutput.MaterialOutputModel import MaterialOutputModel
+from cura.PrinterOutput.NetworkedPrinterOutputDevice import NetworkedPrinterOutputDevice, AuthState
+from cura.PrinterOutput.PrinterOutputModel import PrinterOutputModel
+from plugins.UM3NetworkPrinting.src.UM3PrintJobOutputModel import UM3PrintJobOutputModel
+from .Models import (
+    CloudClusterPrinter, CloudClusterPrintJob, JobUploadRequest, JobUploadResponse, PrintResponse, CloudClusterStatus,
+    CloudClusterPrinterConfigurationMaterial
+)
+
+
+##  The cloud output device is a network output device that works remotely but has limited functionality.
+#   Currently it only supports viewing the printer and print job status and adding a new job to the queue.
+#   As such, those methods have been implemented here.
+#   Note that this device represents a single remote cluster, not a list of multiple clusters.
+#
+#   TODO: figure our how the QML interface for the cluster networking should operate with this limited functionality.
+class CloudOutputDevice(NetworkedPrinterOutputDevice):
```

Comment:
Why not make this inherit from the Cluster UM3 output and change the things you actually want different? Right now there is quite a bit of code duplication.

Reply:
We're still figuring out whether to do that or not depending on how much we need to override if we use cluster output device as base.

Reply:
Fair enough. It's not that big of an issue, but i want to prevent having duplicated code as much as possible. We might want to add an extra level of abstraction though (eg; move the notion that there are multiple printers to a shared cloud / cura connect parent)

Reply:
Yup, that's part actually of what we're refactoring right now :)